### PR TITLE
[WIP] Add GPML support

### PR DESCRIPTION
**Pull request by MridulS**

### CODE REVIEW

File: graphs/util/gpml_util_interface.py [hunk 1]
```diff
@@ -0,0 +1,345 @@
...
+    labels = tree.findall(GPML+'Label')
+    G, parse_json, count = parse_labels(G, parse_json, count, labels)
+    # This is a crude version of getting interactions (edges) working
+    # Look for all the interactions
+    interactions = tree.findall(GPML+'Interaction')
+    G, parse_json, count = parse_interactions(G, parse_json, count, interactions)
+    print G.nodes(data=True)
+    layout = set_gpml_layout(G, parse_json)
+
+    # To add nodes to the group we parse through the group
```
Comment (tmmurali) [inline graphs/util/gpml_util_interface.py]: Change the comment to "To add nodes to the groups to which they belong, we parse the group ..."

File: graphs/util/gpml_util_interface.py [hunk 2]
```diff
@@ -0,0 +1,345 @@
...
+    metadata = {'description':'example1','title':'Example 1 Graph','tags':[]}
+    out = interface.convertNXToDict(G, metadata=metadata)
+
+    return out, parse_json, layout, title
+
+
+def parse_shapes(G, parse_json, count, shapes):
+    # Create a list of details of all the shape
+    shape_details = []
+    for shape in shapes:
```
Comment (tmmurali) [inline graphs/util/gpml_util_interface.py]: Are we calling them details or attributes? Is "detail" a concept used in GPML or is "attribute" the right word?

File: graphs/util/gpml_util.py [hunk 3]
```diff
@@ -0,0 +1,284 @@
...
+    # No tags or description since CYJS doesn't give me any
+    parse_json['metadata']['tags'] = []
+    parse_json['metadata']['description'] = ""
+    title = title or parse_json['metadata']['name']
+    return parse_json, layout, title
+
+
+def parse_shapes(parse_json, count, shapes):
+    # Create a list of details of all the shape
+    shape_details = []
```
Comment (tmmurali) [inline graphs/util/gpml_util.py]: Is "Details" a term used in GPML? If not, is "attributes" the right term. Use the right term here so tha tyou do not invent new terminology.

File: graphs/util/gpml_util_interface.py [hunk 4]
```diff
@@ -0,0 +1,296 @@
...
+        temp_node = {}
+        temp_node.update(shape.attrib)
+        for more_details in shape:
+            temp_node.update(more_details.attrib)
+        shape_details.append(temp_node)
+    # Add shape elemets to the JSON structure parse_json
+    for node in shape_details:
+        temp_node = {'data': {}}
```
Comment (tmmurali) [inline graphs/util/gpml_util_interface.py]: Use GraphId as the unique id for this shape. Do not use count anymore. Of course, you should ensure that all GraphIds over the graph are unique, just out of caution.

File: graphs/util/gpml_util_interface.py [hunk 5]
```diff
@@ -0,0 +1,296 @@
...
+        interface.add_node(G, count)
+        _node_graphics(G, count, node)
+
+        # Shape element may or may not contain a TextLabel field
+        if 'TextLabel' in node:
+            interface.add_node_label(G, count, node['TextLabel'])
+
+        # Deal with the ShapeType field of the shape element,
+        # improve this, make it more general, too many hacks right now.
+        if 'ShapeType' in node:
```
Comment (tmmurali) [inline graphs/util/gpml_util_interface.py]: Create a dict whose keys are GPML shapes and whose values are Cytoscape.js shapes. Add a function that will map from GPML shape to Cytoscape.js shape and throws an error if the GPML shape is not recognised. 

Document this translation scheme on the help pages.

File: graphs/util/gpml_util_interface.py [hunk 6]
```diff
@@ -0,0 +1,296 @@
...
+                else:
+                    interface.add_node_shape(G, count, 'roundrectangle')
+        else:
+            # if no ShapeType given use rectangle by default.
+            interface.add_node_shape(G, count, 'rectangle')
+        # this is the twist out here, if you compare this with node color
+        # attribute, we use the color given for border_color rather than
+        # color (i.e. text color)
+        if 'Color' in node:
+            temp_node['data']['border_color'] = '#' + node['Color']
```
Comment (MridulS) [inline graphs/util/gpml_util_interface.py]: interface

File: graphs/util/gpml_util_interface.py [hunk 7]
```diff
@@ -0,0 +1,296 @@
...
+
+    
+    metadata = {'description':'','title':  graph_gpml.attrib['Name'] or "GPML",'tags':[]}
+    title = title or metadata['title']
+    parse_json = interface.convertNXToDict(G, metadata=metadata)
+
+    return parse_json, layout, title
+
+
+def parse_shapes(G, count,  graph_gpml):
```
Comment (tmmurali) [inline graphs/util/gpml_util_interface.py]: All methods need API documentation.

File: graphs/util/gpml_util_interface.py [hunk 8]
```diff
@@ -0,0 +1,296 @@
...
+    # the various nodes of the graph
+    G, count = parse_datanodes(G, count, graph_gpml)
+    # print G.node
+    # Parsing the GPML file for Shape elements
+    G, count = parse_labels(G, count, graph_gpml)
+
+
+    # This is a crude version of getting interactions (edges) working
+    # Look for all the interactions
+    interactions =  graph_gpml.findall(GPML+'Interaction')
```
Comment (tmmurali) [inline graphs/util/gpml_util_interface.py]: Move this line into parse_interactions().

File: graphs/util/gpml_util_interface.py [hunk 9]
```diff
@@ -0,0 +1,296 @@
...
+    node_details = []
+    for node in datanodes:
+        temp = {}
+        temp = node.attrib
+        for details in node:
+            temp.update(details.attrib)
+        node_details.append(temp)
+    # Go through the node details of the node
+    for node in node_details:
+        temp_node = {'data': {}}
```
Comment (tmmurali) [inline graphs/util/gpml_util_interface.py]: Parse the GraphID of this node.

File: graphs/util/gpml_util_interface.py [hunk 10]
```diff
@@ -0,0 +1,296 @@
...
+def parse_interactions(G,   count, interactions):
+    # Create a list of (x, y) coordinates of an edge
+    # every element of the list can be either of length of 2 or 3
+    # as curved edges are represented by 3 coordinates.
+    edge_details = []
+    for edge in interactions:
+        for details in edge:
+            temp_edge = []
+            for more in details:
+                if 'X' in more.attrib and 'Y' in more.attrib:
```
Comment (tmmurali) [inline graphs/util/gpml_util_interface.py]: You will need to parse the GraphRefs that correspond to the nodes connected by this interaction.

File: graphs/util/gpml_util_interface.py [hunk 11]
```diff
@@ -0,0 +1,296 @@
...
+    default_layout = json.dumps(default_layout)
+    return default_layout
+
+
+def parse_groups(G, groups):
+    for group in groups:
+        temp_node = {'data': {}}
+        interface.add_node(G, group.attrib['GroupId'])
+        interface.add_node_color(G,group.attrib['GroupId'], 'white')
+        interface.add_node_border_color(G, group.attrib['GroupId'], 'black')
```
Comment (tmmurali) [inline graphs/util/gpml_util_interface.py]: Add this comment: We explicitly make the border of a group to be black so that we can see the group in GraphSpace.

File: graphs/util/gpml_util_interface.py [hunk 12]
```diff
@@ -0,0 +1,296 @@
...
+    return G
+
+
+def _node_graphics(G, count, node):
```
Comment (tmmurali) [inline graphs/util/gpml_util_interface.py]: Call this function _update_node_graphics().

File: graphs/util/gpml_util_interface.py [hunk 13]
```diff
@@ -0,0 +1,348 @@
...
+    for node in node_attributes:
+        # Give an id to the node element.
+        interface.add_node(G, node['GraphId'], shape='rectangle', border_width=2)
+        # Parse through the 
+        G = update_node_graphics(G, node['GraphId'], node)
+
+        if 'Error' in G:
+            return G
+
+        # if 'Color' in node:
```
Comment (adbharadwaj) [inline graphs/util/gpml_util_interface.py]: Either add an explanation why we need this commented code or delete it.

**Comment by tmmurali** [comment]
Say the following in the description of the pull request:

This pull request adds the ability for the user to upload GPML files through the GraphSpace web interface, e.g., http://graphspace.org/graphs/upload. GPML (Graphical Pathway Markup Language) is the format used by WikiPathways and PathVisio. (hyperlink these words to the WikiPathways and PathVisio home pages). 

The main function is upload_gpml_file() in graphs/util/db.py. Most of the functionality that supports this function is in graphs/util/gpml_util_interface.py. 

upload_gpml_file() follows the [specification of the GPML format](http://www.pathvisio.org/data/gpml/GPML2013a.xsd.html). It rejects GPML files that do not follow this format, e.g., files that do not use GraphIds for nodes and GraphRefs to specify which nodes an interaction connects. Therefore, the only x- and y- coordinates in the GPML file this function records are for shapes, nodes, and text labels. _It ignores the coordinates for endpoints of interactions._ 

Caveats: Write down all the corner cases here or attributes or shapes in the GPML format that do not work correctly or perfectly.

**Comment by adbharadwaj** [comment]
I am gonna merge this pull request and test it locally. Code looks good so far.
